### PR TITLE
[Test Only]Widen clamp inputs for eltwise integer typecast tests

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### PR Category
Test Only

### Issue
https://github.com/tenstorrent/tt-metal/issues/46237

## Summary

Follow-up to #46118 for remaining **eltwise** integer typecast tests. Widens inputs and uses clamp-aware `eltwise_typecast` goldens on sharded, legacy, and row-major paths that still used in-range inputs or naive `.int()` / `.to(uint8)` references.

Also extracts shared helpers used by `test_run_eltwise_typecast_op` and the updated sharded/RM tests.

**Note:** Moving helpers to `typecast_test_helpers.py` also widens float→integer input ranges in `test_run_eltwise_typecast_op` (e.g. `fp32→int32`, `bf16→uint8` now use `[-1000, 80000]` / `[-512, 512]` instead of default `0..100`). Integer-input bounds are unchanged vs #46118.

### `typecast_test_helpers.py` (new)

* `typecast_test_input_bounds()` — derives `(low, high)` from output dtype limits plus margin:
  * signed/float → `uint8`: ±2×(max+1) (wrap/truncation)
  * signed/float → `uint16`/`uint32`/`int32`: below 0 and above 65535 (clamp/saturation)
  * unsigned inputs: `low = 0`; `high` capped by `_UNSIGNED_INPUT_MAX` (e.g. `uint16` → `[0, 65535]`)
* `make_typecast_test_input()` — integer vs float input generation
* `uses_exact_integer_typecast_check()` — integer exact-equality predicate
* `assert_integer_typecast_equal()` — int64 comparison via `assert_equal`

**Note:** Explicit paths from #46118 (`int32/uint32 → uint16`, `int32 → uint8`, etc.) are preserved. Additional integer combos that previously defaulted to `(0, 100)` now get widened bounds where appropriate. Float→integer paths in `test_run_eltwise_typecast_op` also use the derived ranges.

### `test_eltwise_typecast.py`
- Refactor `test_run_eltwise_typecast_op` to shared helpers; float→integer combos now get widened bounds
- Sharded fp32→int32 (4 tests): `[-1000, 80000]` + `eltwise_typecast` golden
- RM bf16→uint8 (2 tests): `[-512, 512]` + wrap-aware golden

### `test_typecast_int.py`
- `test_typecast_subcore_grid`: full uint16 range for `in_data1`; signed negatives in `in_data2`
- `test_typecast_subcore_grid_large_tensor`: full uint16 range

###Pipeline status

- [x] Sanity test
- [x] (Runtime) Sanity Tests
- [x] BHPC

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/typecast_test_2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/typecast_test_2)

### CODE REVIEW

File: tests/ttnn/python_api_testing/typecast_test_helpers.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def uses_exact_integer_typecast_check(tt_input_dtype, tt_output_dtype):
+    if tt_output_dtype in (ttnn.int32, ttnn.uint32, ttnn.uint8):
+        return True
+    return tt_output_dtype == ttnn.uint16 and tt_input_dtype in (
+        ttnn.int32,
+        ttnn.uint16,
+        ttnn.uint32,
+        ttnn.uint8,
+    )
```
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply tests/ttnn/python_api_testing/typecast_test_helpers.py:111]: Intentional — float/bf8/bf4 → uint16 keeps PCC because device rounding can differ from the Python golden by ±1 (assert_equal would false-fail). This PR targets integer-input clamp/saturation paths (#46035). Documented in the helper docstring

**Review by Aswinmcw — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: tests/ttnn/python_api_testing/typecast_test_helpers.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+FLOAT_INPUT_DTYPES = {ttnn.bfloat16, ttnn.float32, ttnn.bfloat8_b, ttnn.bfloat4_b}
+
+
+def typecast_test_input_bounds(tt_input_dtype, tt_output_dtype):
+    """Out-of-range inputs for integer typecast: clamp (→uint16), wrap (→uint8).
+
+    int32 inputs may be negative; uint16/uint32 inputs stay >= 0 to avoid reinterpretation
+    on device (e.g. -1 as uint32 → 4294967295) that would mismatch clamp-aware golden.
+    """
+    if tt_output_dtype == ttnn.uint16 and tt_input_dtype == ttnn.int32:
```
Comment (amahmudTT) [inline tests/ttnn/python_api_testing/typecast_test_helpers.py]: Nit : Upto you if you want to address this. Instead of hardcoding, the logic can be filtered out. And the ranges defined by the limit of the data types. 


For example 

- if input type allows negative 

             if output allows negative 
                   if abs (min val of input) > abs (min val of output) 

                               low = -2 * min val of output 
              else 
                              low = - max val of output 
........

Comment (VirdhatchaniKN) [reply tests/ttnn/python_api_testing/typecast_test_helpers.py]: Addressed - typecast_test_input_bounds() now derives ranges from output dtype min/max plus margin (2× span for uint8 wrap, +14465 above uint16 max for clamp). Unsigned inputs stay ≥ 0, and high is capped by input dtype max via _UNSIGNED_INPUT_MAX.

File: tests/ttnn/python_api_testing/typecast_test_helpers.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    ttnn.uint16: 65535,
+    # uint32 inputs are stored as int32 in tests; clamp_high fits without wrap.
+    ttnn.uint32: 2**31 - 1,
+}
+# Representable limits per dtype (uint32 output capped to uint16-scale for practical randint).
+_DTYPE_MIN = {
+    ttnn.uint8: 0,
+    ttnn.uint16: 0,
+    ttnn.uint32: 0,
+    ttnn.int32: -(2**31),
```
Comment (amahmudTT) [inline tests/ttnn/python_api_testing/typecast_test_helpers.py]: should this be 32 ?

File: tests/ttnn/python_api_testing/typecast_test_helpers.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    elif tt_output_dtype in (ttnn.uint16, ttnn.uint32, ttnn.int32):
+        high = _CLAMP_HIGH
+    else:
+        high = out_max + 100
+
+    # Unsigned inputs cannot represent values above their dtype max (wrap on device).
+    if tt_input_dtype in _UNSIGNED_INPUT_DTYPES:
+        high = min(high, _UNSIGNED_INPUT_MAX[tt_input_dtype])
+
+    return low, high
```
Comment (amahmudTT) [inline tests/ttnn/python_api_testing/typecast_test_helpers.py:88]: did you do some printout and check that this function is working as you want ?

File: tests/ttnn/unit_tests/operations/eltwise/test_typecast_int.py [hunk 5]
```diff
@@ -93,8 +93,8 @@ def test_typecast_uint16(device):
 def test_typecast_subcore_grid(device, shape, sub_core_grid):
     torch.manual_seed(0)
 
-    in_data1 = torch.randint(0, 65500, (shape), dtype=torch.int32)
-    in_data2 = torch.randint(0, 128000, (shape), dtype=torch.int32)
+    in_data1 = torch.randint(0, 65536, (shape), dtype=torch.int32)
```
Comment (amahmudTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_typecast_int.py]: Now that you have defined the ranges of the data types, will it be possible to root out all the hard coded numbers like these ?

File: tests/ttnn/python_api_testing/typecast_test_helpers.py (hunk 3 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/python_api_testing/typecast_test_helpers.py]: -(2**31) is the representable minimum for signed 32-bit (-2^(32-1). The exponent is 31, not 32). Test low is capped at _CLAMP_LOW_MAGNITUDE (-1000), so we never generate full-range int32 values. Renamed to _INT32_MIN / _INT32_MAX for clarity.

File: tests/ttnn/python_api_testing/typecast_test_helpers.py (hunk 4 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/python_api_testing/typecast_test_helpers.py:88]: Yes Anil. I printed the bounds for the integer input/output pairs we care about — unsigned inputs stay at low=0, and high is capped at whatever the input dtype can actually represent. I also generated sample tensors and looked at min/max; for example on uint16→uint32 everything stays ≤65535, so there's no wrap mismatch with the golden.

For the clamp paths, int32→uint16 does produce negatives and values above 65535, and the golden correctly clamps to [0, 65535].

I also added test_typecast_test_input_bounds in test_eltwise_typecast.py so we keep this covered in the eltwise sanity CI runs.

File: tests/ttnn/unit_tests/operations/eltwise/test_typecast_int.py (hunk 5 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/eltwise/test_typecast_int.py]: Done - test_typecast_subcore_grid and test_typecast_subcore_grid_large_tensor now use typecast_test_input_bounds() + make_typecast_test_input() instead of hardcoded 65536 / 128001. Fixed-value tests (test_typecast_uint16, etc.) keep explicit tensors since they target specific corner values.

**Review by amahmudTT — APPROVED** [review]
